### PR TITLE
Allow to create REST Client directly through rest.Config

### PR DESCRIPTION
**Pull request by ingvagabund**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:

Currently, `k8s.io/client-go/kubernetes.New` function allows to create a clientset
with pre-created REST client. However, the REST client passed into individual clientset
clients (e.g. AppsV1, BatchV1, ...) has fixed `GroupVersion` and does not take into account
`GroupVersion` of other clientset clients. The REST client (among other things) carries serializer
configuration for a single `GroupVersion`. So, only a single clientset client initialized
with the `GroupVersion` can be actually used to perform serialization. Any other clientset client
will complain about impossible conversion from its `GroupVersion` to the only `GroupVersion`.
Which makes `k8s.io/client-go/kubernetes.New` function generally unusable for serialization.
    
Documentation for `rest.ContentConfig.GroupVersion` field says:
```
// GroupVersion is the API version to talk to. Must be provided when initializing
// a RESTClient directly. When initializing a Client, will be set with the default
// code version.
```
However, k8s.io/client-go/kubernetes.New call does not lead to invocation of any
defaulting mechanism.
    
Instead of invoking function New, I suggest to invoke `NewForConfig`, resp. `NewForConfigOrDie`
functions with `rest.Config` containing new public field `RESTClientBuilder`. Both functions lead
to invocation of `GroupVersion` client specific `setConfigDefaults` function that sets the right
`GroupVersion` field. Followed by invocation of a function stored under `RESTClientBuilder` which
allows to decide which `GroupVersion` is actually requested and thus build the REST client with
the right `GroupVersion`-ed serializer. If `RESTClientBuilder` field is not set,
it defaults to the original `rest.RESTClientFor` function.
    
Example of usage:
```go
...
import (
	"k8s.io/client-go/kubernetes"
	...
)
...
kubernetes.NewForConfigOrDie(&rest.Config{
	RESTClientBuilder: func(config *rest.Config) (rest.Interface, error) {
		switch config.GroupVersion.String() {
		case "v1":
			return coreRestClient, nil
		case "policy/v1beta1":
			return policyRestClient, nil
		case "storage.k8s.io/v1":
			return storageRestClient, nil
		case "apps/v1":
			return appsRestClient, nil
		default:
			return nil, fmt.Errorf("unknown GroupVersion: %v", config.GroupVersion.String())
		}
	},
```

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
`k8s.io/client-go/rest.Config` data type extended with `RESTClient()` method and `RESTClientBuilder` field to provide control over building REST client for `GroupVersion`-ed clientset clients.

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/rest/config.go [hunk 1]
```diff
@@ -131,6 +131,11 @@ type Config struct {
 	// Version forces a specific version to be used (if registered)
 	// Do we need this?
 	// Version string
+
+	// RESTClientBuilder allows to inject function for building REST client.
+	// E.g. for mocking REST client to be used by various clientset clients.
+	// If not set, Config.RESTClient() falls to the default RESTClientFor.
+	RESTClientBuilder func(config *Config) (Interface, error)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/config.go]: It's unclear to me how this solves the issue in the PR description. Without demonstrated use in the repo, I question whether we should add this injection capability.

File: staging/src/k8s.io/client-go/rest/config.go [hunk 2]
```diff
@@ -147,6 +152,17 @@ func (sanitizedAuthConfigPersister) String() string {
 	return "rest.AuthProviderConfigPersister(--- REDACTED ---)"
 }
 
+// RESTClient returns a RESTClient that satisfies the requested attributes on a client Config
+// object. Note that a RESTClient may require fields that are optional when initializing a Client.
+// A RESTClient created by this method is generic - it expects to operate on an API that follows
+// the Kubernetes conventions, but may not be the Kubernetes API.
+func (c *Config) RESTClient() (Interface, error) {
```
Comment (jpbetz) [inline staging/src/k8s.io/client-go/rest/config.go]: This increases the responsibility of the configuration substantially; it is now client builder. It is more tightly coupled with `rest.Interface`. I'm not sure it's worth doing just for the minor convenience gained.

**Comment by ingvagabund** [comment]
Thanks for the review. I found a better way and completely dropped my mocked REST client. The fake clientset already implements watching for events so I can go one layer up and completely rely on the clientset.

/close

**Pull request closed without merging**